osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -403,6 +403,32 @@ protected override BeatmapSetInfo CreateModel(ArchiveReader reader)
                 return null;
             }
 
+            // pre-check to make sure that the archive hasn't already been imported.
+            bool anyToImport = false;
+
+            foreach (var file in reader.Filenames.Where(f => f.EndsWith(".osu", StringComparison.OrdinalIgnoreCase)))
```

Comment:
Note that this doesn't consider non `.osu` files in the archive. I don't think this a big deal, but checking all files reduces performance substantially even though it's still much faster than current master.